extensions/amp-story/1.0/amp-story-grid-layer.js

```diff
@@ -108,6 +108,43 @@ export class AmpStoryGridLayer extends AmpStoryBaseLayer {
     return this.prerenderAllowed_;
   }
 
+  // TODO(26866): switch back to onMeasuredChange and remove the layoutBox
+  // equality checks.
+  /** @override */
+  onLayoutMeasure() {
```

Comment:
Given that there are multiple grid-layer per story-page, and multiple story-page per story, could we move this code to [here](https://github.com/ampproject/amphtml/blob/master/extensions/amp-story/1.0/amp-story-page.js#L506), so it only runs on the first story-page?

It would require running a query like `querySelectorAll('amp-story-grid-layer[aspect-ratio]')` outside of the scope of the said story-page, which is rather hacky, but this code would run only once vs dozens of times with the current implementation.

Reply:
The way we defined this compromise is per layer for now. Because not all layers follow the letterboxing. For instance, background layer can be (and recommended to be) fullbleed. However, we are not yet ready to standardize how that should happen. Hence this compromise of per-layer aspect ratio.

Reply:
Right, let's keep it per layer, but still move the code to the `amp-story-page` so it measures only once, and loops through all the layers to set the the font-size based on the aspect-ratio value they provided.

Reply:
For example, the [latest story from WaPo](https://www.washingtonpost.com/graphics/2020/politics/amp-stories/super-tuesday-primary-states-delegate-search/) has 110 grid-layers, and `onLayoutMeasure` runs pretty often (which is why we have the `getLayoutBox` check on the amp-story-page implementation).

Reply:
Hmm. A bit confused. You want to move the code to the `amp-story-page`, but applying per layer? I.e. if the `aspect-ratio` is specified on a layer, then different layers could have different values, right? Maybe, instead, we should do the "only do if changed" check?

Reply:
Sorry, let me clarify.

The issues I'd like to work around are:
1. For a regular story, this code will likely run several hundreds of time. Stories have many grid-layers (110 for that story linked above), and `onLayoutMeasure` triggers several times.
2. Using `parent.clientWidth/Height`, even in a measure context, would be safer to get rid of

We already have a place where we measure the width/height of a page/grid-layer, to set some CSS variables: [here](https://github.com/ampproject/amphtml/blob/master/extensions/amp-story/1.0/amp-story-page.js#L506).

We want to only measure once, and benefit from the runtime cached `layoutBox` values as much as possible (the code we have already handles some story specific edge cases here).

#### Option A
I understand that the code is per layer, but something like this would work. I know that it is rather hacky, but to me, the performance and code consolidation improvements are really worth it.

```
// in amp-story-page.onLayoutMeasure()
const storyEl = this.element.parentElement;
const width = w; // Do something with the existing code getting the width.
const height = h; // Do something with the existing code getting the height.

const layers = storyEl.querySelectorAll('amp-story-grid-layer[aspect-ratio]');

// loop through the grid layers, and set new scoped CSS variables on each one of them
```

#### Option B
Another solution where `amp-story-page` does not have to update code outside of its scope would be to:
1. Use the `amp-story-store-service` to store the width/height of a story-page/grid-layer
2. grid-layers stop relying on `onLayoutMeasure`, and instead subscribe to this state, and update their CSS variables on change

This solution is actually probably better than the first one. The idea being to re-use the width/height that we already compute, and only run the code when needed.
Happy to show you an example of how this `store-service` can be used, but this is extremely straightforward ([example](https://github.com/ampproject/amphtml/pull/18780/files)).

Reply:
Cool. I much rather like Option B. Will reimplement using that option.

Reply:
Done. PTAL.